rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/BundleApi.java

```diff
@@ -0,0 +1,158 @@
+        @QueryParam("force") @DefaultValue("false")
+        Boolean force);
+
+    @POST
+    @Consumes({MediaType.APPLICATION_JSON, "application/x-yaml",
+        // see http://stackoverflow.com/questions/332129/yaml-mime-type
+        "text/yaml", "text/x-yaml", "application/yaml"})
+    @ApiOperation(
+            value = "Adds types to the registry from a given BOM YAML/JSON descriptor (creating a bundle with just this file in it)",
+            response = BundleInstallationRestResult.class
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Error processing the given YAML"),
+            @ApiResponse(code = 201, message = "Items added successfully")
+    })
+    public Response createFromYaml(
+            @ApiParam(name = "yaml", value = "BOM YAML declaring the types to be installed", required = true)
+            @Valid String yaml,
+            @ApiParam(name="force", value="Force installation including replacing any different bundle of the same name and version")
+            @QueryParam("force") @DefaultValue("false")
+            Boolean forceUpdate);
+
+    @POST
+    @Consumes({"application/x-zip", "application/x-jar"})
+    @ApiOperation(
+            value = "Adds types to the registry from a given JAR or ZIP",
+            notes = "Accepts either an OSGi bundle JAR, or ZIP which will be turned into bundle JAR. Either format must "
+                    + "contain a catalog.bom at the root of the archive, which must contain the bundle and version key.",
+            response = BundleInstallationRestResult.class)
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Error processing the given archive, or the catalog.bom is invalid"),
+            @ApiResponse(code = 201, message = "Catalog items added successfully")
+    })
+    public Response createFromArchive(
+            @ApiParam(
+                    name = "archive",
+                    value = "Bundle to install, in ZIP or JAR format, requiring catalog.bom containing bundle name and version",
+                    required = true)
+            byte[] archive,
+            @ApiParam(name = "force", value = "Whether to forcibly remove it, even if in use and/or errors", required = false, defaultValue = "false")
+            @QueryParam("force") @DefaultValue("false")
+            Boolean force);
+
+    @POST
+    @Consumes // anything (if doesn't match other methods with specific content types
+    @ApiOperation(
+            value = "Adds types to the registry from the given item, autodetecting type as ZIP/JAR or BOM YAML",
+            response = BundleInstallationRestResult.class
+    )
+    @ApiResponses(value = {
+            @ApiResponse(code = 400, message = "Error processing the given archive, or the catalog.bom is invalid"),
+            @ApiResponse(code = 201, message = "Catalog items added successfully")
+    })
+    public Response createAutodetecting(
```

Comment:
I lean towards not doing this - instead rely on `Content-Type: ...` being supplied.

I say this because of the annoying behaviour in the existing api for things like `POST /v1/catalog`. If you post an invalid .bom file, it first tries to parse it as a .bom which fails, and then tries to parse it as a different format. The error that gets returned is often misleading (e.g. it gives you details of the second attempt, rather than the first attempt). I've hit this problem while trying to help real newbee (but smart) users of Brooklyn.

Looking at this new code, it would be quite easy to lose the error message that tells you your yaml was invalid, with it instead telling you that it's not a zip file.

What does this really buy us? The UI and CLI will always supply it. If using `curl` then you have to type a little bit more, but I think that's fine.

Reply:
this is all being removed as we will just have `POST /catalog` -- but i agree with you